lib/aws/rails/action_mailbox/sns_notification.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require 'aws-sdk-sns'
+
+module Aws
+  module Rails
+    module ActionMailbox
+      class SnsNotification
+        class MessageContentError < StandardError; end
+
+        def initialize(request_body)
+          @request_body = request_body
+        end
+
+        def subscription_confirmed?
+          confirmation_response.code&.start_with?('2')
+        end
+
+        def verified?
+          Aws::SNS::MessageVerifier.new.authentic?(@request_body)
+        end
+
+        def topic
+          notification.fetch(:TopicArn)
+        end
+
+        def type
+          notification.fetch(:Type)
+        end
+
+        def message_content
+          raise MessageContentError, 'Incoming emails must have notificationType `Received`' unless receipt?
+
+          if content_in_s3?
+            s3_content
+          else
+            return message[:content] unless destination
+
+            "X-Original-To: #{destination}\n#{message[:content]}"
+          end
+        end
+
+        private
+
+        def notification
+          @notification ||= JSON.parse(@request_body, symbolize_names: true)
+        rescue JSON::ParserError => e
+          Rails.logger.warn("Unable to parse SNS notification: #{e}")
+          nil
+        end
+
+        def s3_content
+          require 'aws-sdk-s3'
+
+          Aws::S3::Client
```

Comment:
Well, since this is called from a controller I think even if it's memoized it'd still be regenerated unless it's set at boot time, right, and then if it's set at boot time it would be instantiating for those not using it 🤔

Reply:
added the framework from my understanding 🤔